### DISCUSSION TITLE
Signing MacOS Binaries / App Store distribution

### DISCUSSION THREAD

**Discussion by emlynmac** [jamulussoftware/jamulus#1851] [General]
Hi,

As I understand, we don't have a signing method for jamulus currently.
I'm a macOS / iOS developer by trade and have the developer membership and I'd be willing to make the signing / App Store releases for jamulus moving forward if there is interest.

**Comment by ann0see** [comment]
Hi @emlynmac Thanks for volunteering!
Yes, Jamulus is currently not signed and [iOS support](https://github.com/jamulussoftware/jamulus/pull/1512) is only in a very early stage. Personally, I'd really appreciate having a signed version of Jamulus! A few questions would arise on the [signing topic ](https://github.com/jamulussoftware/jamulus/discussions/1042):

How would this affect our release process? Would you need to compile, sign and upload Jamulus on your own? Could we integrate signing in our CI process? How can we ensure that you're releasing the official version (without malware, etc)?

**Reply by emlynmac** [reply]
Oh, sorry I should have been clearer. 
macOS signing initially - there's way more people using that.

I could also do iOS too.

In terms of the signing process, for iOS there are ways to re-sign exported IPA files if you already have that as part of your build exports. I believe the same of the macOS system, but I need to verify that. 

I'll start by making some corresponding projects on my dev account and will test that out. Ultimately, the certificates could be wrapped into the CI process. How are you currently doing that?

**Reply by emlynmac** [reply]
Looks like someone has already registered the app ID net.sourceforge.llcon.Jamulus in the App Store. I'll be using io.jamulus.app for now.

**Reply by ann0see** [reply]
Yes, iOS is in an early stage and will probably not pass the verification process. (No need to worry about that now since macOS is more important).

The CI is run on basis of this file via GitHub actions: https://github.com/jamulussoftware/jamulus/blob/master/.github/workflows/autobuild.yml

Although I'm not sure how the CI (which we call "Autobuild" in Jamulus's terms) currently works, since a lot of changes have been made to it in the last months, it should work as follows:

1. Get Qt and all the dependencies
2. Build it via macdeployqt (and probably xcodebuild?): See: https://github.com/jamulussoftware/jamulus/blob/master/mac/deploy_mac.sh which I've only scanned
3. Package the app via dmgbuild
4. Upload the .dmg file to either a release or artifact on GitHub

**Reply by emlynmac** [reply]
I'll set up the same environment here on a different Mac and see how it all fits together.

**Reply by emlynmac** [reply]
```
codesign --verify --verbose=4 jamulus_3.8.0_mac.dmg 
jamulus_3.8.0_mac.dmg: valid on disk
jamulus_3.8.0_mac.dmg: satisfies its Designated Requirement
```
So in theory I have signed the existing binary DMG file from the download section - is there a way I can get you the file to test out? It's 24mb or so.
Local testing suggests I might have to go down the notarization route though, so that would mean a custom build I think as there needs to be entitlements set.

**Reply by ann0see** [reply]
Thanks for doing this!
I think you could upload it as GitHub release or use Disroot Upload or similar: https://upload.disroot.org/

**Reply by emlynmac** [reply]
OK - a resigned version of the release is here: https://upload.disroot.org/r/TB6uNsKL#kiETlHC1+6NxL8KxhatEiDah7ojDLNJYZVALomRLEtM=
I don't expect this to work...

A fresh build (I found my old Catalina laptop!) with the adjusted bundle ID and signing baked into the process is here:
https://upload.disroot.org/r/at5TALGu#WY9kNhglrFIh7Q1T2rsRpgp6Dm8lsQEO4AqblZ+h0AM=

**Reply by ann0see** [reply]
Thanks. I‘ll test it soon!

 I just found this article: https://localazy.com/blog/how-to-automatically-sign-macos-apps-using-github-actions

but I haven’t tried or read it completely.

**Reply by ann0see** [reply]
Actually on Big Sur both versions now just seem to warn that the app was downloaded from the web. It now no longer says that it’s unverified. 

Good Job!

Personally, I would prefer a manual signing process which means more work for you but eliminates the need of sharing secrets. We‘d only need to sign release/release candidate executables, nothing in beta/nightly stage.

**Reply by emlynmac** [reply]
Yes, I think I'd prefer that too - I can just do that once the releases come up and there's no pushing the certificates up anywhere online which I'm _much_ happier about.

The simplest way will be to resign the .dmg, which is a one-liner:
```
codesign --deep --force --verify --sign "Developer ID Application: Bolton Technology Consulting Ltd. (ENJ2T4A3FZ)" jamulus_3.8.0_mac.dmg
```

So the next thing I'm going to look at is how to get it to the Mac app store if that's ok with you. This has a few more hoops, which will likely need a custom build in order to get it all working properly with hardened runtime etc, but I'll figure that out.
Let me know if you're NOT ok with this plan.

**Reply by ann0see** [reply]
I just checked the Jamulus Server. It doesn’t seem to be signed (?) maybe someone else with a Mac should try to install it. Also the _CodeSignature folder seems to be missing on the Jamulus.app file. 

I _think_ nobody can stop you from putting the app on the Apple App store. Especially if all contributors are happy. I personally am, but Jamulus has been developed by multiple people.
There might be legal questions we should first clarify (GPL).

**Reply by emlynmac** [reply]
oh right - no, I forgot about that... I'll add that...
EDIT: actually, they should both have been signed. I wonder why that happened.

Yes, OK, let's do that first. There's usually no issue with GPL per se, it's more about having to adhere to their guidelines etc and be ok with that, risk not being allowed etc.

**Reply by ann0see** [reply]
I think we should include some statement (like the Nextcloud client does for their iOS version). @jamulussoftware/maindevelopers I think we should give a thought on that.

**Reply by ngocdh** [reply]
I think we just need to state that Jamulus is released under GPL 2.0.

Google site:apps.apple.com gpl 2.0

@emlynmac would it be possible for you to publish the iOS app on the App Store, or at least on TestFlight? Is it a lot of work?

**Reply by emlynmac** [reply]
If you get another chance, can you test this package?
https://upload.disroot.org/r/V_CmgSwb#9l6Ns/AQQm/GTKEX24whtdiCddkzc8b0RY0NTP+mCYE=
I changed the identifiers a little that should make both apps work.

**Reply by emlynmac** [reply]
@ngocdh Yes, it's possible. How much work, I don't yet know...

For QT 5, this means using Xcode 11 (downloading for my old laptop).  I'll have to figure out if I can use the free tools otherwise it's a paid sub to QT tools, which looks kinda expensive.

**Reply by ann0see** [reply]
I think FLOSS projects like Jamulus is can use the free version of Qt

**Reply by ngocdh** [reply]
> If you get another chance, can you test this package?
> https://upload.disroot.org/r/V_CmgSwb#9l6Ns/AQQm/GTKEX24whtdiCddkzc8b0RY0NTP+mCYE=
> I changed the identifiers a little that should make both apps work.

"“jamulus_3.8.0_mac-2.dmg” cannot be opened because it is from an unidentified developer."

Is it normal?

**Reply by ngocdh** [reply]
> I think FLOSS projects like Jamulus is can use the free version of Qt

```
Open Source Usage Obligations  
The majority of the Qt modules are available under the LGPL v3 and GPL v3 open source license. When developing under this license your obligations are to:

Provide a re-linking mechanism for Qt libraries
Provide a license copy & explicitly acknowledge Qt use
Make a Qt source code copy available for customers
Accept that Qt source code modifications are non-proprietary
Make “open” consumer devices
Accept Digital Rights Management terms, please see the GPL FAQ
Take special consideration when attempting to enforce software patents
```

Should be good.

**Reply by ann0see** [reply]
Your new signed .dmg now seems to be blocked. Maybe the test I made was successful because I already had a trusted version installed.

**Reply by emlynmac** [reply]
Yeah, this one is weird. The certs are trusted on my Big Sur Mac, but not on the Catalina one. I'll try regenerating on the Catalina Mac and see....

**Reply by emlynmac** [reply]
Can you try installing this: https://www.apple.com/certificateauthority/AppleWWDRCAG3.cer
It is the required missing certificate.

**Reply by ann0see** [reply]
Actually, I am running Big Sur.

**Reply by emlynmac** [reply]
So to clarify.
* This version https://upload.disroot.org/r/TB6uNsKL#kiETlHC1+6NxL8KxhatEiDah7ojDLNJYZVALomRLEtM= worked (resigned only)
* The rest have not since.
* You have not installed the apple intermediate certificate?

**Reply by ann0see** [reply]
Resigned .dmg:

Server and Client bundles don‘t have a _CodeSignature folder. The client could be opened without a warning; the server didn’t open. Maybe this is related to the fact that I already had a version of the Jamulus Client installed and verified

Client which you compiled: seems to have a valid signature

Last link you sent: gives a warning before the .dmg can even open. 

Therefore I think that only the version you compiled yourself works correctly. 

The certificate has not been installed yet

**Reply by emlynmac** [reply]
OK, try this: https://upload.disroot.org/r/THKB3Hoa#iDopo7m5NVhWYuixhYeBlWGQko3NJjCwhtdMDggMUC8=
https://upload.disroot.org/r/THKB3Hoa#iDopo7m5NVhWYuixhYeBlWGQko3NJjCwhtdMDggMUC8=
Passed notarization with apple so should be good to go.
One quick caveat.... It won't be able to connect to audio. 
After a bit of digging, I need to rework the `macdeployqt` script to allow adding of entitlements, which will mean some hacking and then push back to the qt repo.
I wasn't expecting this to be so time consuming! Should have something final by the weekend....

**Reply by ann0see** [reply]
Can try it soon. So you‘re saying that this needs changes in the Qt source? Are you sure it’s not possible to add configuration parameters which set these entitlements?

**Reply by emlynmac** [reply]
After looking through the `macdeployqt` source, it looks for an entitlements file in the bundle. Now I've put that in place, the `codesign` finds it and uses it.

**Reply by ann0see** [reply]
Thanks for both links. Yes, the app seems to be signed now but audio isn’t working as you said. Can the entitlements file be added by qmake?

**Reply by emlynmac** [reply]
yes, the entitlements can be added - and found it seems.
So I have now a fully working / signed version of 3.8.0
https://upload.disroot.org/r/RkFDZ7hH#Ln/WujuCs0Czokpxt44ZKfx/d/2vtNaEfFmkAgTCqZM=

This does require making some changes to the build files and having the signing certificate handy, so I'll have to rebuild locally when it is release time.

I'll make a PR for the changes

**Reply by emlynmac** [reply]
PR: https://github.com/jamulussoftware/jamulus/pull/1856

**Comment by ngocdh** [comment]
Thank you @emlynmac. I would love to see Jamulus on the App Store one day :)

**Reply by emlynmac** [reply]
Me too - let's see if I can make that happen....

**Reply by emlynmac** [reply]
I'm almost there - I have built a version that works, let's see if apple approve it for use.

**Comment by ann0see** [comment]
I see two next actions here:

1. Clarification of the App Store release with our "explanation"/"disclaimer" (can be delayed)
2. Signed macOS binaries (this is finished technically; however we haven’t clarified how @emlynmac can show that his binaries are trustworthy and only based on the code from this repo.) I don’t think there’s a need (yet) to include in into the CI process.

**Reply by gene96817** [reply]
For establishing the binaries published by @emlynmtac are trustworthy, a simple low-tech way is that we publish on jamulus.io the signature of the trusted code. Then if anyone cares, they can check that the code they receive through the Apple Store has the right signature.

**Reply by ann0see** [reply]
I doubt that’s possible since he needs to compile with a signing certificate —> hash gets invalid. Same with a normal non AppStore signed version of Jamulus

**Comment by gene96817** [comment]
1- Just take a convenient signature or hash.
2- posting the hash on Jamulus.io is a statement that we recognize that distribution.
3- provide instructions of how to validate (check) the hash. Some users know they should check but they don't know how.

**Comment by emlynmac** [comment]
I’ll have to set up some sort of signing from a branch in my fork of GitHub. I need to build with my signing certificate and then notarized with Apple

**Reply by ann0see** [reply]
Ok. So GH compiles it, we get the binaries (verify that the code is not different) and then upload it as official release?

**Comment by danryu** [comment]
Hi guys,

Not sure where you are at with this, but I managed to put some time into this recently.
After a _lot_ of trial and error (and battling with the sparse Apple dev cli tool documentation) I finally have a working autobuild that generates:
- signed macOS pkg file for submission to MacOS App Store
- signed iOS ipa file for submission to iOS App Store

These files can then be submitted directly to the App Store via eg Transporter on macOS, and will appear within the app store submission build choices. Easy!

I built on the previous work by @ann0see  and @emlynmac  - please see changes in our fork at https://github.com/koord-live/koord-realtime.
most changes being in:
https://github.com/koord-live/koord-realtime/blob/kdev/ios/deploy_ios.sh
https://github.com/koord-live/koord-realtime/blob/kdev/.github/autobuild/ios.sh

_Note: I had to move to macOS 11 to build for iOS, in order to use XCode 13.2.1 (13+ being necessary for SDK compatibility for App Store upload). Still with Qt 5.15.2 - I tried 6.0, 6.1 and 6.2 but with vague build failures in each case._

If there is interest in a PR I would be happy to do this - I have just merged from upstream master so it will be an easy merge.

Requirements:

1) Set up the following resources in in https://developer.apple.com/account/resources/certificates/list 

Certificates:
- Mac Installer Distribution 
- Mac App Distribution 
- iOS Distribution 

Identifier:
- app ID (bundleID)

Profile:
- iOS Distribution (AppStore) - ProvisioningProfile

2) Add certs + profile to Github Secrets as per
https://docs.github.com/en/actions/deployment/deploying-xcode-applications/installing-an-apple-certificate-on-macos-runners-for-xcode-development

3) Run the autobuild :)
Signed IPA and PKG files are produced.

**Reply by emlynmac** [reply]
@danryu thanks for your work. 
We currently have a means to push the signed macOS builds out via gitHub auto build on my fork - it would be nice to get the App Store sorted out too.

Can you make a PR for your work and I can look at it? 

Please can you separate the steps into one for macOS App Store and one for iOS App Store?

Thanks!

**Reply by danryu** [reply]
Yes no problem, that should be straightforward.

**Reply by danryu** [reply]
For clarity I created 2 separate PRs:
- macOS signed for App Store: https://github.com/jamulussoftware/jamulus/pull/2624
- iOS signed for App Store: https://github.com/jamulussoftware/jamulus/pull/2625

The end product files in each case should pass all verification checks when uploading to App Store via eg Transporter.

**Reply by ann0see** [reply]
Great! Thank you very much. I think most of this is up to emlyn since he's the person distributing the signed app?
